server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java

```diff
@@ -21,214 +21,170 @@
+        }
+    }
+
+    private void doBootstrap(VotingConfiguration votingConfiguration) {
+        assert transportService.getLocalNode().isMasterNode();
 
-                transportService.getThreadPool().scheduleUnlessShuttingDown(unconfiguredBootstrapTimeout, Names.SAME, new Runnable() {
+        try {
+            votingConfigurationConsumer.accept(votingConfiguration);
+        } catch (Exception e) {
+            logger.warn(new ParameterizedMessage("exception when bootstrapping with {}, rescheduling", votingConfiguration), e);
+            transportService.getThreadPool().scheduleUnlessShuttingDown(TimeValue.timeValueSeconds(10), Names.GENERIC,
+                new Runnable() {
                     @Override
                     public void run() {
-                        // TODO: remove the following line once schedule method properly preserves thread context
-                        threadContext.markAsSystemContext();
-                        final GetDiscoveredNodesRequest request = new GetDiscoveredNodesRequest();
-                        logger.trace("sending {}", request);
-                        transportService.sendRequest(transportService.getLocalNode(), GetDiscoveredNodesAction.NAME, request,
-                            new TransportResponseHandler<GetDiscoveredNodesResponse>() {
-                                @Override
-                                public void handleResponse(GetDiscoveredNodesResponse response) {
-                                    logger.debug("discovered {}, starting to bootstrap", response.getNodes());
-                                    awaitBootstrap(response.getBootstrapConfiguration());
-                                }
-
-                                @Override
-                                public void handleException(TransportException exp) {
-                                    final Throwable rootCause = exp.getRootCause();
-                                    if (rootCause instanceof ClusterAlreadyBootstrappedException) {
-                                        logger.debug(rootCause.getMessage(), rootCause);
-                                    } else {
-                                        logger.warn("discovery attempt failed", exp);
-                                    }
-                                }
-
-                                @Override
-                                public String executor() {
-                                    return Names.SAME;
-                                }
-
-                                @Override
-                                public GetDiscoveredNodesResponse read(StreamInput in) throws IOException {
-                                    return new GetDiscoveredNodesResponse(in);
-                                }
-                            });
+                        doBootstrap(votingConfiguration);
```

Comment:
Is it so important to use the same voting configuration for retrying? As long as it matches the requirements it should be good, no? In case where it's an automated best-effort bootstrap, we have no strong guarantees and keeping the same voting config does not matter either?

Reply:
Probably not, but this is how we've modelled it. Retrying here is a fairly weird thing to be doing anyway, requiring one of the following:

- the node is not bootstrapped, and also not a `CANDIDATE`
- the node is no longer in touch with a quorum of peers
- an I/O error occurred

I'm in two minds about whether to retry automatically at all, as opposed to simply logging the issue and letting the operator restart the node if they really need to.

Reply:
I thought about this some more and concluded that retrying was not a particularly helpful thing to do on an exception, so as of 2ff94a06747d034cebef26f1eb27d33754c43bd0 we no longer do so.

Reply:
I think we should have retries, for the reason of making bootstrapping fault-tolerant. It would be bad if in a 3 node cluster, one node would go down during bootstrapping, and none of the 2 other nodes would complete the bootstrapping. Yet this is a possibilty, with node 1 going down and node 2 and 3 both use node 1 and themselves as initial voting config (because node 1 was shortly available during discovery).
Similarly I feel that a node turning to follower (because of follower-check, so does not have accepted a cluster state yet) is not a good reason to abort setting initial config. Given that we're not changing the cluster state anymore when setting initial config, I think we should change the check from (not be CANDIDATE) to (not have a voting configuration).

Reply:
Ok, I reverted that commit in 9d2bc207a07951a5f968f8d34a9ddfe027a9b705.

We already only bootstrap if we do (not have a voting configuration) - this check is in addition to the check on the mode. And yet we do change the cluster state when setting the initial config, so I don't follow the reasoning. Do you mean to say that we don't change the cluster state version on bootstrapping?

Reply:
> Do you mean to say that we don't change the cluster state version on bootstrapping?

yes.

The mode check might be superfluous.

Reply:
Note that this comment is still unaddressed:

> I think we should have retries, for the reason of making bootstrapping fault-tolerant. It would be bad if in a 3 node cluster, one node would go down during bootstrapping, and none of the 2 other nodes would complete the bootstrapping. Yet this is a possibilty, with node 1 going down and node 2 and 3 both use node 1 and themselves as initial voting config (because node 1 was shortly available during discovery).

I feel that this is an important situation to cover and we should also have CoordinatorTests for it, i.e., start 3-node cluster with initial_master_nodes set to all nodes, do a little runRandomly, then completely isolate one of the nodes, and see if the other 2 can form a cluster (stabilization).

Reply:
I removed the mode check in f30ab6e6b8c0551b66a63a7ce7e2f3a63e9f942e and we agreed to follow up on the potential availability issue later since it requires some careful thought.